File: contrib-drivers-master/FirebaseIoTApp/src/main/java/com/example/androidthings/myproject/MainActivity.java
/*
 * Copyright 2016, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.androidthings.myproject;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import android.view.View;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.io.IOException;
import java.net.URISyntaxException;
import java.util.Scanner;
//fezhat board
import com.google.android.things.contrib.driver.fezhat.FezHat;
import com.google.android.things.contrib.driver.fezhat.Akselerasi;
import com.google.android.things.contrib.driver.fezhat.Color;
import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.PeripheralManagerService;

import com.google.gson.Gson;
import java.text.DecimalFormat;


import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.media.Image;
import android.media.ImageReader;
import android.os.Build;
import android.os.HandlerThread;
import android.speech.tts.TextToSpeech;
import android.speech.tts.UtteranceProgressListener;
import android.support.annotation.NonNull;
import android.view.KeyEvent;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.Toast;

import com.example.androidthings.myproject.classifier.Classifier;
import com.example.androidthings.myproject.classifier.TensorFlowImageClassifier;

//bisa dibuang
import com.google.android.things.pio.Gpio;
import com.google.android.things.pio.PeripheralManagerService;

import java.util.List;
import java.util.Locale;
import java.util.concurrent.atomic.AtomicBoolean;

import static android.Manifest.permission.CAMERA;
import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;

import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;


/**
 * Skeleton of the main Android Things activity. Implement your device's logic
 * in this class.
 *
 * Android Things peripheral APIs are accessible through the class
 * PeripheralManagerService. For example, the snippet below will open a GPIO pin and
 * set it to HIGH:
 *
 * <pre>{@code
 * PeripheralManagerService service = new PeripheralManagerService();
 * mLedGpio = service.openGpio("BCM6");
 * mLedGpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);
 * mLedGpio.setValue(true);
 * }</pre>
 *
 * For more complex peripherals, look for an existing user-space driver, or implement one if none
 * is available.
 *
 */

public class MainActivity extends Activity implements ImageReader.OnImageAvailableListener {
    private static final String TAG = MainActivity.class.getSimpleName();

    private static final int INTERVAL_BETWEEN_BLINKS_MS = 500;
    private Handler mHandler = new Handler();

    private FezHat hat;
    private boolean next;
    private int i;

    protected TextView TxtStatus;
    protected TextView TxtTemp;
    protected TextView TxtLight;
    protected Button BtnReceive;
    protected TextView TxtAccel;

    //tensorflow app
    private static final int PERMISSIONS_REQUEST = 1;

    private ImagePreprocessor mImagePreprocessor;
    private TextToSpeech mTtsEngine;
    private TtsSpeaker mTtsSpeaker;
    private CameraHandler mCameraHandler;
    private TensorFlowImageClassifier mTensorFlowClassifier;

    private HandlerThread mBackgroundThread;
    private Handler mBackgroundHandler;

    private ImageView mImage;
    private TextView[] mResultViews;

    private AtomicBoolean mReady = new AtomicBoolean(false);

    private FirebaseDatabase database;
    private DatabaseReference myRef;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_layout);
        if (hasPermission()) {
            if (savedInstanceState == null) {
                init();
            }
        } else {
            requestPermission();
        }

    }

    protected void Setup() throws URISyntaxException, IOException {
        TxtAccel = (TextView) findViewById(R.id.txtAccel);
        TxtLight = (TextView) findViewById(R.id.txtLight);
        TxtTemp = (TextView) findViewById(R.id.txtTemp);
        mImage = (ImageView) findViewById(R.id.imageView);
        mResultViews = new TextView[3];
        mResultViews[0] = (TextView) findViewById(R.id.result1);
        mResultViews[1] = (TextView) findViewById(R.id.result2);
        mResultViews[2] = (TextView) findViewById(R.id.result3);

        /*
        TxtStatus = (TextView) findViewById(R.id.txtStatus);
        BtnReceive = (Button) findViewById(R.id.btnReceive);
        BtnReceive.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                try {
                    ReceiveData();
                } catch (IOException io) {

                } catch (URISyntaxException es) {

                }
            }
        });
        */
        this.hat = FezHat.Create();
        this.hat.S1.SetLimits(500, 2400, 0, 180);
        this.hat.S2.SetLimits(500, 2400, 0, 180);
        //set as not ready
        this.hat.D2.setColor(Color.Red());
        this.hat.D3.setColor(Color.Red());

    }

    public void SendMessage(SensorData data) throws URISyntaxException, IOException {
        Gson gson = new Gson();
        String msgStr = gson.toJson(data);
        try {
            myRef.setValue(msgStr);

        } catch (Exception e) {
        }

    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        try {
            if (mBackgroundThread != null) mBackgroundThread.quit();
        } catch (Throwable t) {
            // close quietly
        }
        mBackgroundThread = null;
        mBackgroundHandler = null;

        try {
            if (mCameraHandler != null) mCameraHandler.shutDown();
        } catch (Throwable t) {
            // close quietly
        }
        try {
            if (mTensorFlowClassifier != null) mTensorFlowClassifier.close();
        } catch (Throwable t) {
            // close quietly
        }

        if (mTtsEngine != null) {
            mTtsEngine.stop();
            mTtsEngine.shutdown();
        }
        Log.d(TAG, "onDestroy");
        mHandler.removeCallbacks(mBlinkRunnable);


    }

    // Permission-related methods. This is not needed for Android Things, where permissions are
    // automatically granted. However, it is kept here in case the developer needs to test on a
    // regular Android device.
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[],
                                           @NonNull int[] grantResults) {
        switch (requestCode) {
            case PERMISSIONS_REQUEST: {
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED
                        && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
                    init();
                } else {
                    requestPermission();
                }
            }
        }
    }

    private boolean hasPermission() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        return checkSelfPermission(CAMERA) == PackageManager.PERMISSION_GRANTED
                && checkSelfPermission(WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
    }

    private void requestPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (shouldShowRequestPermissionRationale(CAMERA) ||
                    shouldShowRequestPermissionRationale(WRITE_EXTERNAL_STORAGE)) {
                Toast.makeText(this, "Camera AND storage permission are required for this demo",
                        Toast.LENGTH_LONG).show();
            }
            requestPermissions(new String[]{CAMERA, WRITE_EXTERNAL_STORAGE}, PERMISSIONS_REQUEST);
        }
    }


    private void init() {
        //init firebase
        database = FirebaseDatabase.getInstance();
        myRef = database.getReference("telemetry");
        //received data
        // Read from the database
        myRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                // This method is called once with the initial value and again
                // whenever data at this location is updated.
                String value = dataSnapshot.getValue(String.class);
                Log.d(TAG, "Value is: " + value);
            }

            @Override
            public void onCancelled(DatabaseError error) {
                // Failed to read value
                Log.w(TAG, "Failed to read value.", error.toException());
            }
        });

        //init peripheral
        try {
            PeripheralManagerService service = new PeripheralManagerService();
            Log.d(TAG, "Available GPIO: " + service.getGpioList());
            Log.d(TAG, "onCreate");
            try {
                Setup();
            } catch (URISyntaxException ex) {


            }
        } catch (IOException e) {
            Log.d(TAG, e.getMessage());
        }
        mHandler.post(mBlinkRunnable);
        //run background process for TF
        mBackgroundThread = new HandlerThread("BackgroundThread");
        mBackgroundThread.start();
        mBackgroundHandler = new Handler(mBackgroundThread.getLooper());
        mBackgroundHandler.post(mInitializeOnBackground);
    }


    private Runnable mInitializeOnBackground = new Runnable() {
        @Override
        public void run() {
            mImagePreprocessor = new ImagePreprocessor(CameraHandler.IMAGE_WIDTH,
                    CameraHandler.IMAGE_HEIGHT, TensorFlowImageClassifier.INPUT_SIZE);

            mTtsSpeaker = new TtsSpeaker();
            mTtsSpeaker.setHasSenseOfHumor(true);
            mTtsEngine = new TextToSpeech(MainActivity.this,
                    new TextToSpeech.OnInitListener() {
                        @Override
                        public void onInit(int status) {
                            if (status == TextToSpeech.SUCCESS) {
                                mTtsEngine.setLanguage(Locale.US);
                                mTtsEngine.setOnUtteranceProgressListener(utteranceListener);
                                mTtsSpeaker.speakReady(mTtsEngine);
                            } else {
                                Log.w(TAG, "Could not open TTS Engine (onInit status=" + status
                                        + "). Ignoring text to speech");
                                mTtsEngine = null;
                            }
                        }
                    });
            mCameraHandler = CameraHandler.getInstance();
            mCameraHandler.initializeCamera(
                    MainActivity.this, mBackgroundHandler,
                    MainActivity.this);

            mTensorFlowClassifier = new TensorFlowImageClassifier(MainActivity.this);

            setReady(true);
        }
    };

    private Runnable mBackgroundClickHandler = new Runnable() {
        @Override
        public void run() {
            if (mTtsEngine != null) {
                mTtsSpeaker.speakShutterSound(mTtsEngine);
            }
            mCameraHandler.takePicture();
        }
    };

    private UtteranceProgressListener utteranceListener = new UtteranceProgressListener() {
        @Override
        public void onStart(String utteranceId) {
            setReady(false);
        }

        @Override
        public void onDone(String utteranceId) {
            setReady(true);
        }

        @Override
        public void onError(String utteranceId) {
            setReady(true);
        }
    };

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        Log.d(TAG, "Received key down: " + keyCode + ". Ready = " + mReady.get());
        if (keyCode == KeyEvent.KEYCODE_ENTER || keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {

            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    private void setReady(boolean ready) {
        mReady.set(ready);
            try {
                if(ready){
                    hat.D2.setColor(Color.Green());
                    hat.D3.setColor(Color.Green());
                }else{
                    hat.D2.setColor(Color.Yellow());
                    hat.D3.setColor(Color.Yellow());
                }


            } catch (IOException e) {
                Log.w(TAG, "Could not set LED", e);
            }
    }

    private void TakePicture(){
        if (mReady.get()) {
            setReady(false);
            mBackgroundHandler.post(mBackgroundClickHandler);
        } else {
            Log.i(TAG, "Sorry, processing hasn't finished. Try again in a few seconds");
        }
    }

    @Override
    public void onImageAvailable(ImageReader reader) {
        final Bitmap bitmap;
        try (Image image = reader.acquireNextImage()) {
            bitmap = mImagePreprocessor.preprocessImage(image);
        }

        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mImage.setImageBitmap(bitmap);
            }
        });

        final List<Classifier.Recognition> results = mTensorFlowClassifier.recognizeImage(bitmap);

        Log.d(TAG, "Got the following results from Tensorflow: " + results);
        if (mTtsEngine != null) {
            // speak out loud the result of the image recognition
            mTtsSpeaker.speakResults(mTtsEngine, results);
        } else {
            // if theres no TTS, we don't need to wait until the utterance is spoken, so we set
            // to ready right away.
            setReady(true);
        }

        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                for (int i = 0; i < mResultViews.length; i++) {
                    if (results.size() > i) {
                        Classifier.Recognition r = results.get(i);
                        mResultViews[i].setText(r.toString());
                    } else {
                        mResultViews[i].setText(null);
                    }
                }
            }
        });
    }

    protected Runnable mBlinkRunnable = new Runnable() {

        @Override
        public void run() {
            try {
                mHandler.postDelayed(mBlinkRunnable, INTERVAL_BETWEEN_BLINKS_MS);
                DecimalFormat formatter = new DecimalFormat("######.000");
                SensorData data = new SensorData();
                Akselerasi accel = hat.GetAcceleration();
                data.Light = hat.GetLightLevel();
                data.Temp = hat.GetTemperature();
                data.Accelleration = "X:" + String.valueOf(accel.X) + " Y: " + String.valueOf(accel.Y) + " Z:" + String.valueOf(accel.Z);
                String lightStr = String.valueOf(data.Light);
                String Temp = formatter.format(data.Temp);
                String AccelStr = data.Accelleration;
                String Btn18Str = String.valueOf(hat.IsDIO18Pressed());
                String Btn22Str = String.valueOf(hat.IsDIO22Pressed());
                String AnalogStr = String.valueOf(hat.ReadAnalog(FezHat.AnalogPin.Ain1));
                TxtLight.setText(lightStr + " Lux");
                TxtAccel.setText(AccelStr);
                TxtTemp.setText(Temp + " C");
                Log.e(TAG, "Light:" + lightStr);
                Log.e(TAG, "Temp:" + Temp);
                Log.e(TAG, "Acceleration:" + AccelStr);
                if(hat.IsDIO18Pressed() || hat.IsDIO22Pressed()){
                    TakePicture();
                }
                try {
                    SendMessage(data);
                }catch (IOException ex){

                }catch (URISyntaxException ex){

                }
                /*
                if ((i++ % 5) == 0) {
                    String LedsTextBox = String.valueOf(next);

                    hat.setDIO24On(next);
                    hat.D2.setColor(next ? Color.Green() : Color.Black());
                    hat.D3.setColor(next ? Color.Green() : Color.Black());

                    hat.WriteDigital(FezHat.DigitalPin.DIO16, next);
                    hat.WriteDigital(FezHat.DigitalPin.DIO26, next);

                    hat.SetPwmDutyCycle(FezHat.PwmPin.Pwm5, next ? 1.0 : 0.0);
                    hat.SetPwmDutyCycle(FezHat.PwmPin.Pwm6, next ? 1.0 : 0.0);
                    hat.SetPwmDutyCycle(FezHat.PwmPin.Pwm7, next ? 1.0 : 0.0);
                    hat.SetPwmDutyCycle(FezHat.PwmPin.Pwm11, next ? 1.0 : 0.0);
                    hat.SetPwmDutyCycle(FezHat.PwmPin.Pwm12, next ? 1.0 : 0.0);

                    next = !next;
                }

                if (hat.IsDIO18Pressed()) {
                    hat.S1.setPosition(hat.S1.getPosition() + 5.0);
                    hat.S2.setPosition(hat.S2.getPosition() + 5.0);

                    if (hat.S1.getPosition() >= 180.0) {
                        hat.S1.setPosition(0.0);
                        hat.S2.setPosition(0.0);
                    }
                }

                if (hat.IsDIO22Pressed()) {
                    if (hat.MotorA.getSpeed() == 0.0) {
                        hat.MotorA.setSpeed(0.7);
                        hat.MotorB.setSpeed(-0.7);
                    }
                } else {
                    if (hat.MotorA.getSpeed() != 0.0) {
                        hat.MotorA.setSpeed(0.0);
                        hat.MotorB.setSpeed(0.0);
                    }
                }
                */
            } catch (IOException e) {
                Log.e(TAG, "Error on Jalan", e);

            }
        }
    };

}

/*
public class ImageClassifierActivity extends Activity implements ImageReader.OnImageAvailableListener {

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        setContentView(R.layout.activity_camera);

    }


    @Override
    protected void onDestroy() {
        super.onDestroy();

    }


}*/

File: contrib-drivers-master/FirstApp/src/main/java/com/example/androidthings/myproject/MainActivity.kt
/*
 * Copyright 2016, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.androidthings.myproject

import android.app.Activity
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.widget.Button
import android.widget.TextView
import android.view.View

import java.text.SimpleDateFormat
import java.util.Calendar
import java.io.IOException
import java.net.URISyntaxException
import java.util.Scanner

//fezhat board
import com.google.android.things.contrib.driver.fezhat.FezHat
import com.google.android.things.contrib.driver.fezhat.Akselerasi
import com.google.android.things.contrib.driver.fezhat.Color
import com.google.android.things.pio.Gpio
import com.google.android.things.pio.PeripheralManagerService
import com.google.android.things.pio.UartDevice
import com.google.android.things.pio.UartDeviceCallback

//azure lib
import com.microsoft.azure.iothub.DeviceClient
import com.microsoft.azure.iothub.IotHubClientProtocol
import com.microsoft.azure.iothub.IotHubEventCallback
import com.microsoft.azure.iothub.IotHubMessageResult
import com.microsoft.azure.iothub.IotHubStatusCode
import com.microsoft.azure.iothub.Message
import com.microsoft.azure.iothub.MessageCallback

//mqtt

import com.google.gson.Gson
import org.eclipse.paho.client.mqttv3.*
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence
import java.text.DecimalFormat

import com.google.gson.JsonDeserializer
import com.google.gson.JsonSyntaxException


/**
 * Skeleton of the main Android Things activity. Implement your device's logic
 * in this class.
 *
 * Android Things peripheral APIs are accessible through the class
 * PeripheralManagerService. For example, the snippet below will open a GPIO pin and
 * set it to HIGH:
 *
 * <pre>`PeripheralManagerService service = new PeripheralManagerService();
 * mLedGpio = service.openGpio("BCM6");
 * mLedGpio.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW);
 * mLedGpio.setValue(true);
`</pre> *
 *
 * For more complex peripherals, look for an existing user-space driver, or implement one if none
 * is available.
 *
 */
class MainActivity : Activity() {
    internal var connString = "HostName=FreeDeviceHub.azure-devices.net;DeviceId=AndroidThingsDevice;SharedAccessKey=U5mH0B3POECuZlTCmMA9xEIql1+S+ASQBWOPG1d3iBY="

    private var client: DeviceClient? = null
    private val mHandler = Handler()

    private var hat: FezHat? = null
    private var next: Boolean = false
    private var i: Int = 0

    protected var TxtStatus: TextView? = null
    protected var TxtTemp: TextView? = null
    protected var TxtLight: TextView? = null
    protected var BtnReceive: Button? = null
    protected var TxtAccel: TextView? = null

    // UART Device Name
    private var UART_DEVICE_NAME : String = "USB1-1.4:1.0"//"USB1-1.5:1.0"
    private var mDevice : UartDevice? = null
    var datastr : String = ""

    val publishTopic = "mifmasterz/gdgbogor/data"
    private var mqttAndroidClient : MqttClient? = null;

    protected var mBlinkRunnable: Runnable = object : Runnable {

        override fun run() {
            try {
                mHandler.postDelayed(this, INTERVAL_BETWEEN_BLINKS_MS.toLong())
                val formatter = DecimalFormat("######.000")
                val data = SensorData()
                val accel = hat!!.GetAcceleration()
                data.Light = hat!!.GetLightLevel()
                data.Temp = hat!!.GetTemperature()
                data.Accelleration = "X:" + accel.X.toString() + " Y: " + accel.Y.toString() + " Z:" + accel.Z.toString()
                val lightStr = data.Light.toString()
                val Temp = formatter.format(data.Temp)
                val AccelStr = data.Accelleration
                val Btn18Str = hat!!.IsDIO18Pressed().toString()
                val Btn22Str = hat!!.IsDIO22Pressed().toString()
                val AnalogStr = hat!!.ReadAnalog(FezHat.AnalogPin.Ain1).toString()
                TxtLight?.text = "Light: " + lightStr
                TxtAccel?.text = "Acceleration: " + AccelStr
                TxtTemp?.text = "Temp: $Temp C"
                Log.e(TAG, "Light:" + lightStr)
                Log.e(TAG, "Temp:" + Temp)
                Log.e(TAG, "Acceleration:" + AccelStr)
                Log.e(TAG, "Counter:" + i)
                Log.e(TAG, "Next:" + next)



                if (i++ % 5 == 0) {
                    val LedsTextBox = next.toString()

                    hat!!.diO24On = next
                    hat!!.D2.color = if (next) Color.Green() else Color.Black()
                    hat!!.D3.color = if (next) Color.Green() else Color.Black()

                    hat!!.WriteDigital(FezHat.DigitalPin.DIO16, next)
                    hat!!.WriteDigital(FezHat.DigitalPin.DIO26, next)

                    hat!!.SetPwmDutyCycle(FezHat.PwmPin.Pwm5, if (next) 1.0 else 0.0)
                    hat!!.SetPwmDutyCycle(FezHat.PwmPin.Pwm6, if (next) 1.0 else 0.0)
                    hat!!.SetPwmDutyCycle(FezHat.PwmPin.Pwm7, if (next) 1.0 else 0.0)
                    hat!!.SetPwmDutyCycle(FezHat.PwmPin.Pwm11, if (next) 1.0 else 0.0)
                    hat!!.SetPwmDutyCycle(FezHat.PwmPin.Pwm12, if (next) 1.0 else 0.0)

                    next = !next
                }

                if (hat!!.IsDIO18Pressed()) {
                    hat!!.S1.position = hat!!.S1.position + 5.0
                    hat!!.S2.position = hat!!.S2.position + 5.0

                    if (hat!!.S1.position >= 180.0) {
                        hat!!.S1.position = 0.0
                        hat!!.S2.position = 0.0
                    }
                }

                if (hat!!.IsDIO22Pressed()) {
                    if (hat!!.MotorA.speed == 0.0) {
                        hat!!.MotorA.speed = 0.7
                        hat!!.MotorB.speed = -0.7
                    }
                } else {
                    if (hat!!.MotorA.speed != 0.0) {
                        hat!!.MotorA.speed = 0.0
                        hat!!.MotorB.speed = 0.0
                    }
                }
                Log.d(TAG, "jalan lagi...")
                //sending data to azure

                try {
                    SendMessage(data)
                    Log.d(TAG, "data has been pushed to azure...")
                    val timeStamp = SimpleDateFormat("dd MM yy HH:mm:ss").format(Calendar.getInstance().time)
                    TxtStatus?.text = "Data has been pushed to azure at " + timeStamp
                } catch (ex: URISyntaxException) {

                } finally {

                }
            } catch (e: IOException) {
                Log.e(TAG, "Error on Jalan", e)

            }

        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)
        try {
            val service = PeripheralManagerService()
            Log.d(TAG, "Available GPIO: " + service.gpioList)
            Log.d(TAG, "onCreate")
            try {
                Setup()
            } catch (ex: URISyntaxException) {

            }

            mHandler.post(mBlinkRunnable)
        } catch (e: IOException) {
            Log.d(TAG, e.message)
        }

    }

    @Throws(URISyntaxException::class, IOException::class)
    protected fun Setup() {
        TxtAccel = findViewById(R.id.txtAccel) as TextView
        TxtLight = findViewById(R.id.txtLight) as TextView
        TxtTemp = findViewById(R.id.txtTemp) as TextView
        TxtStatus = findViewById(R.id.txtStatus) as TextView
        BtnReceive = findViewById(R.id.btnReceive) as Button
        BtnReceive?.setOnClickListener {
            try {
                ReceiveData()
            } catch (io: IOException) {

            } catch (es: URISyntaxException) {

            }
        }
        this.hat = FezHat.Create()
        this.hat!!.S1.SetLimits(500, 2400, 0.0, 180.0)
        this.hat!!.S2.SetLimits(500, 2400, 0.0, 180.0)
        // Comment/uncomment from lines below to use HTTPS or MQTT protocol
        // IotHubClientProtocol protocol = IotHubClientProtocol.HTTPS;
        val protocol = IotHubClientProtocol.MQTT

        client = DeviceClient(connString, protocol)

        try {
            client!!.open()
        } catch (e1: IOException) {
            println("Exception while opening IoTHub connection: " + e1.toString())
        } catch (e2: Exception) {
            println("Exception while opening IoTHub connection: " + e2.toString())
        }

        //USB1-1.5:1.0
        try {
            val manager = PeripheralManagerService()


            var deviceList : List<String> = manager.getUartDeviceList()
            if (deviceList.isEmpty()) {
                Log.i(TAG, "No UART port available on this device.");
            } else {
                Log.i(TAG, "List of available devices: " + deviceList);
            }


            mDevice = manager.openUartDevice(UART_DEVICE_NAME)
            setFlowControlEnabled(mDevice!!,false)
            configureUartFrame(mDevice!!)
            mDevice?.registerUartDeviceCallback(mUartCallback)
        } catch (e: IOException) {
            Log.w(TAG, "Unable to access UART device", e)
        }
        try {
            mqttAndroidClient = MqttClient("tcp://cloud.makestro.com:1883", "androidthing-device",MemoryPersistence())
            mqttAndroidClient?.setCallback(object : MqttCallback {

                override fun connectionLost(cause: Throwable) {
                    Log.w(TAG, "The Connection was lost.")
                }

                @Throws(Exception::class)
                override fun messageArrived(topic: String, message: MqttMessage) {
                    Log.w(TAG,"Incoming message: " + String(message.payload))
                }

                override fun deliveryComplete(token: IMqttDeliveryToken) {

                }
            })
            val mqttConnectOptions = MqttConnectOptions()
            mqttConnectOptions.userName = "mifmasterz"
            mqttConnectOptions.password = "123qweasd".toCharArray()

            mqttAndroidClient?.connect(mqttConnectOptions);

            mqttAndroidClient?.subscribe(publishTopic)


        } catch ( e:MqttException) {
            e.printStackTrace();
        }

    }

    fun publishMessage(msg:String) {

        try {
            val message = MqttMessage()
            message.payload = msg.toByteArray()
            mqttAndroidClient?.publish(publishTopic, message)
            Log.w(TAG, "Message Published")
            /*
            if (mqttAndroidClient?.isConnected()==false) {
                addToHistory(mqttAndroidClient?.getBufferedMessageCount().toString() + " messages in buffer.")
            }*/
        } catch (e: MqttException) {
            System.err.println("Error Publishing: " + e.message)
            e.printStackTrace()
        }

    }

    @Throws(IOException::class)
    fun writeUartData(uart:UartDevice,messagestr:String) :Unit {
        val newline = System.getProperty("line.separator")
        var message=messagestr+newline
        var buffer:ByteArray = message.toByteArray()
        var count = uart.write(buffer, buffer.size);
        Log.d(TAG, "Wrote " + count + " bytes to peripheral");
    }

    private val mUartCallback = object : UartDeviceCallback() {
        override fun onUartDeviceDataAvailable(uart: UartDevice): Boolean {
            // Read available data from the UART device
            try {
                readUartBuffer(uart)
            } catch (e: IOException) {
                Log.w(TAG, "Unable to access UART device", e)
            }

            // Continue listening for more interrupts
            return true
        }

        override fun onUartDeviceError(uart: UartDevice, error: Int) {
            Log.w(TAG, uart.toString() + ": Error event " + error)
        }
    }

    @Throws(IOException::class)
    fun setFlowControlEnabled(uart: UartDevice, enable: Boolean) {
        if (enable) {
            // Enable hardware flow control
            uart.setHardwareFlowControl(UartDevice.HW_FLOW_CONTROL_AUTO_RTSCTS)
        } else {
            // Disable flow control
            uart.setHardwareFlowControl(UartDevice.HW_FLOW_CONTROL_NONE)
        }
    }

    @Throws(IOException::class)
    fun readUartBuffer(uart:UartDevice ) {

        val newline = System.getProperty("line.separator")

        // Maximum amount of data to read at one time
        var maxCount : Int = 1000
        var buffer : ByteArray = ByteArray(maxCount)

        var count:Int
        count = uart.read(buffer, buffer.size)
        while (count  > 0) {
            Log.d(TAG, "Read " + count + " bytes from peripheral");
            var reconstitutedString = String(buffer, 0, count);
            reconstitutedString = reconstitutedString
            val hasNewline = reconstitutedString.contains(newline)
            datastr += reconstitutedString
            if(hasNewline) {
                if(datastr.length > 0) {
                    Log.d(TAG, "DATA REC: " + datastr)
                    publishMessage(datastr)
                    try {
                        var gson: Gson = Gson()
                        val data = gson.fromJson(datastr, SensorXbee::class.java)
                        if (data.light < 600) {
                            writeUartData(mDevice!!, "{\"alarm\":\"true\"}");
                        } else {
                            writeUartData(mDevice!!, "{\"alarm\":\"false\"}");
                        }
                    } catch (ex: JsonSyntaxException) {
                        Log.d(TAG,ex.message)
                    }
                }
                datastr = ""
            }
            count = uart.read(buffer, buffer.size)
        }

    }

    @Throws(IOException::class)
    fun configureUartFrame(uart: UartDevice) {
        // Configure the UART port
        uart.setBaudrate(9600)
        uart.setDataSize(8)
        uart.setParity(UartDevice.PARITY_NONE)
        uart.setStopBits(1)
    }

    @Throws(URISyntaxException::class, IOException::class)
    fun ReceiveData() {


        // Comment/uncomment from lines below to use HTTPS or MQTT protocol
        // IotHubClientProtocol protocol = IotHubClientProtocol.HTTPS;
        val protocol = IotHubClientProtocol.MQTT

        val client = DeviceClient(connString, protocol)

        if (protocol == IotHubClientProtocol.MQTT) {
            val callback = MessageCallbackMqtt()
            val counter = Counter(0)
            client.setMessageCallback(callback, counter)
        } else {
            val callback = MessageCallback()
            val counter = Counter(0)
            client.setMessageCallback(callback, counter)
        }

        try {
            client.open()
        } catch (e1: IOException) {
            println("Exception while opening IoTHub connection: " + e1.toString())
        } catch (e2: Exception) {
            println("Exception while opening IoTHub connection: " + e2.toString())
        }

        try {
            Thread.sleep(2000)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        client.close()
    }

    // Our MQTT doesn't support abandon/reject, so we will only display the messaged received
    // from IoTHub and return COMPLETE
    protected class MessageCallbackMqtt : com.microsoft.azure.iothub.MessageCallback {
        override fun execute(msg: Message, context: Any): IotHubMessageResult {
            val counter = context as Counter
            println(
                    "Received message " + counter.toString()
                            + " with content: " + String(msg.bytes, Message.DEFAULT_IOTHUB_MESSAGE_CHARSET))

            counter.increment()

            return IotHubMessageResult.COMPLETE
        }
    }

    protected class MessageCallback : com.microsoft.azure.iothub.MessageCallback {
        override fun execute(msg: Message, context: Any): IotHubMessageResult {
            val counter = context as Counter
            println(
                    "Received message " + counter.toString()
                            + " with content: " + String(msg.bytes, Message.DEFAULT_IOTHUB_MESSAGE_CHARSET))

            val switchVal = counter.get() % 3
            val res: IotHubMessageResult
            when (switchVal) {
                0 -> res = IotHubMessageResult.COMPLETE
                1 -> res = IotHubMessageResult.ABANDON
                2 -> res = IotHubMessageResult.REJECT
                else ->
                    // should never happen.
                    throw IllegalStateException("Invalid message result specified.")
            }

            println("Responding to message " + counter.toString() + " with " + res.name)

            counter.increment()

            return res
        }
    }

    @Throws(URISyntaxException::class, IOException::class)
    fun SendMessage(data: SensorData) {
        val gson = Gson()
        val msgStr = gson.toJson(data)
        try {
            val msg = Message(msgStr)
            msg.setProperty("messageCount", Integer.toString(i))
            println(msgStr)
            val eventCallback = EventCallback()
            client!!.sendEventAsync(msg, eventCallback, i)
        } catch (e: Exception) {
        }

    }

    protected class EventCallback : IotHubEventCallback {
        override fun execute(status: IotHubStatusCode, context: Any) {
            val i = context as Int
            println("IoT Hub responded to message " + i.toString()
                    + " with status " + status.name)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy")
        mHandler.removeCallbacks(mBlinkRunnable)
        if (mDevice != null) {
            try {
                mDevice?.unregisterUartDeviceCallback(mUartCallback)
                mDevice?.close()
                mDevice = null
            } catch (e: IOException) {
                Log.w(TAG, "Unable to close UART device", e)
            }

        }
        try {
            client!!.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    protected class Counter(protected var num: Int) {

        fun get(): Int {
            return this.num
        }

        fun increment() {
            this.num++
        }

        override fun toString(): String {
            return Integer.toString(this.num)
        }
    }

    companion object {
        private val TAG = MainActivity::class.java!!.getSimpleName()

        private val INTERVAL_BETWEEN_BLINKS_MS = 2000
    }
}
